dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/dataentry/DataValueContextController.java

```diff
@@ -0,0 +1,99 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.controller.dataentry;
+
+import static org.hisp.dhis.commons.collection.CollectionUtils.mapToList;
+
+import java.util.List;
+import java.util.Set;
+
+import lombok.AllArgsConstructor;
+
+import org.hisp.dhis.category.CategoryOptionCombo;
+import org.hisp.dhis.common.DhisApiVersion;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.datavalue.DataExportParams;
+import org.hisp.dhis.datavalue.DataValue;
+import org.hisp.dhis.datavalue.DataValueAudit;
+import org.hisp.dhis.datavalue.DataValueAuditService;
+import org.hisp.dhis.datavalue.DataValueService;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.PeriodService;
+import org.hisp.dhis.webapi.controller.datavalue.DataValidator;
+import org.hisp.dhis.webapi.mvc.annotation.ApiVersion;
+import org.hisp.dhis.webapi.webdomain.datavalue.DataValueContextDto;
+import org.hisp.dhis.webapi.webdomain.datavalue.DataValueDtoMapper;
+import org.hisp.dhis.webapi.webdomain.datavalue.DataValueQueryParams;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping( "/dataEntry" )
+@AllArgsConstructor
+@ApiVersion( { DhisApiVersion.DEFAULT, DhisApiVersion.ALL } )
+public class DataValueContextController
+{
+    private final DataValueAuditService dataValueAuditService;
+
+    private final DataValueService dataValueService;
+
+    private final PeriodService periodService;
+
+    private final DataValidator dataValidator;
+
+    @GetMapping( "/dataValueContext" )
+    public DataValueContextDto getChangeLog( DataValueQueryParams params )
+    {
+        DataElement de = dataValidator.getAndValidateDataElement( params.getDe() );
+        Period pe = dataValidator.getAndValidatePeriod( params.getPe() );
+        OrganisationUnit ou = dataValidator.getAndValidateOrganisationUnit( params.getOu() );
+        CategoryOptionCombo co = dataValidator.getAndValidateCategoryOptionCombo( params.getCo() );
+        CategoryOptionCombo ao = dataValidator.getAndValidateAttributeOptionCombo( params.getCc(), params.getCp() );
+
```

Comment:
_This is not critical and could be updated later._ 
I think would be nice if we could have a `DataValueContext` class. Then can do sth like this
```
DataValueContext context = dataValidator.getAndValidate( params );
return DataValueContextService.getContextDto( context );
```

or 

``` 
List<Error> errors = new ArrayList<>();
DataValueContext context = DataValueContextService.getAndValidate( params, error -> errors.add( error) );

if ( errors.isEmpty ) return mapper.toDto( context );

return errors;
```

Reply:
Totally agree @vietnguyen. Can do that in a refactor PR.